python/hail/utils/interval.py

```diff
@@ -0,0 +1,286 @@
+        interval._point_type = point_type
+        interval._start = None
+        interval._end = None
+        interval._include_start = None
+        interval._include_end = None
+        super(Interval, interval).__init__()
+        return interval
+
+    @property
+    def start(self):
+        """Start point of the interval.
+
+        Examples
+        --------
+
+        .. doctest::
+
+            >>> interval.start
+            3
+
+        Returns
+        -------
+        Object with type :meth:`.point_type`
+        """
+        if self._start is None:
+            self._start = self.point_type._convert_to_py(self._jrep.start())
+        return self._start
+
+    @property
+    def end(self):
+        """End point of the interval.
+
+        Examples
+        --------
+
+        .. doctest::
+
+            >>> interval.end
+            6
+
+        Returns
+        -------
+        Object with type :meth:`.point_type`
+        """
+        if self._end is None:
+            self._end = self.point_type._convert_to_py(self._jrep.end())
+        return self._end
+
+    @property
+    def include_start(self):
+        """True if interval is inclusive of start.
+
+        Examples
+        --------
+
+        .. doctest::
+
+            >>> interval.include_start
+            True
+
+        Returns
+        -------
+        :obj:`bool`
+        """
+        if self._include_start is None:
+            self._include_start = self._jrep.includeStart()
+        return self._include_start
+
+    @property
+    def include_end(self):
+        """True if interval is inclusive of end.
+
+        Examples
+        --------
+
+        .. doctest::
+
+            >>> interval.include_end
+            False
+
+        Returns
+        -------
+        :obj:`bool`
+        """
+        if self._include_end is None:
+            self._include_end = self._jrep.includeEnd()
+        return self._include_end
+
+    @property
+    def point_type(self):
+        """Type of each element in the interval.
+
+        Examples
+        --------
+
+        .. doctest::
+
+            >>> interval.point_type
+            tint32
+
+        Returns
+        -------
+        :class:`.Type`
+        """
+        return self._point_type
+
+    def contains(self, value):
+        """True if `value` is contained within the interval.
+
+        Examples
+        --------
+
+        .. doctest::
+
+            >>> interval.contains(5)
+            True
+
+            >>> interval.contains(6)
+            False
+
+        Parameters
+        ----------
+        value :
+            Object with type :meth:`.point_type`.
+
+        Returns
+        -------
+        :obj:`bool`
+        """
+
+        from hail.expr.expression import impute_type
+        value_type = impute_type(value)
+        if value_type != self.point_type:
+            raise TypeError("'value' is incompatible with the interval point type: %s, %s" % (value_type, self.point_type))
```

Comment:
don't use this kind of string formatting. Either use ''.format or the new Python `f" {value_type}, {self.point_type}`

also, wrap type printouts in single quotes.